pkg/oci/oci.go

```diff
@@ -0,0 +1,316 @@
+// Copyright 2023 The Inspektor Gadget authors
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package oci
+
+import (
+	"context"
+	"encoding/json"
+	"errors"
+	"fmt"
+	"io"
+	"os"
+	"path/filepath"
+	"runtime"
+
+	"github.com/docker/cli/cli/config"
+	"github.com/docker/cli/cli/config/configfile"
+	"github.com/docker/distribution/reference"
+	ocispec "github.com/opencontainers/image-spec/specs-go/v1"
+	"github.com/sirupsen/logrus"
+	"oras.land/oras-go/v2"
+	"oras.land/oras-go/v2/content/memory"
+	"oras.land/oras-go/v2/content/oci"
+	"oras.land/oras-go/v2/errdef"
+	"oras.land/oras-go/v2/registry/remote"
+	oras_auth "oras.land/oras-go/v2/registry/remote/auth"
+)
+
+type AuthOptions struct {
+	AuthFile string
+	Insecure bool
+}
+
+var (
+	defaultOciStore = "/var/lib/ig/oci-store"
```

Comment:
When we introduce the client/server architecture with #2078, the client might not run as root anymore. so it might not have access to that directory, and it is currently not configurable.

I think it should be under `$XDG_STATE_HOME/ig/oci-store` (`$HOME/.local/state/ig/oci-store`). With https://github.com/adrg/xdg, we could use `xdg.StateHome` and it would automatically pick up the environment variable `$XDG_STATE_HOME`.

Reply:
Then, if the user just wants to build gadgets and push them (but not execute them), it should be possible to run the server as non-root. It seems to require all root privileges to run clang and to upload some files on a remote server. Or should the user uses `ig` binary as non-root for that?

Reply:
I like the idea about having the ability to run `ig build/push/pull/...` as non-root. Maybe as a middle ground we can try to access `/var/ib/ig/oci-store` first. If that fails we fallback to `$HOME/...` or `$XDG...`. Of course with an appropriate info message in the console

Reply:
I'm not sure about that use case, but consider this one:

```bash
# build image as non-root, image saved in $HOME/foo
$ ig image build mygadget:latest

# run the image as root. 
$ sudo ig run mygadget:latest
mygadget:latest not found in $HOME/....
```

How can we support it if we store images in a user dependent folder? That's why I suggested to use `/var/lib` in https://github.com/inspektor-gadget/inspektor-gadget/pull/1863#discussion_r1282344286.